Downsample/Downsample.cpp

```diff
@@ -0,0 +1,167 @@
+	}
+
+	Eigen::MatrixXf x(pointCloud.height(), pointCloud.width());
+	Eigen::MatrixXf y(pointCloud.height(), pointCloud.width());
+	Eigen::MatrixXf z(pointCloud.height(), pointCloud.width());
+	Eigen::MatrixXi r(pointCloud.height(), pointCloud.width());
+	Eigen::MatrixXi g(pointCloud.height(), pointCloud.width());
+	Eigen::MatrixXi b(pointCloud.height(), pointCloud.width());
+	Eigen::MatrixXf contrast(pointCloud.height(), pointCloud.width());
+
+	for (int i = 0; i < pointCloud.height(); i++)
+	{
+		for (int j = 0; j < pointCloud.width(); j++)
+		{
+			x(i, j) = pointCloud(i, j).x;
+			y(i, j) = pointCloud(i, j).y;
+			z(i, j) = pointCloud(i, j).z;
+			r(i, j) = pointCloud(i, j).red();
+			g(i, j) = pointCloud(i, j).green();
+			b(i, j) = pointCloud(i, j).blue();
+			contrast(i, j) = pointCloud(i, j).contrast;
+		}
+	}
+
+	Eigen::MatrixXi redDownsampled = downsampleandround(r, downsamplingFactor);
+	Eigen::MatrixXi greenDownsampled = downsampleandround(g, downsamplingFactor);
+	Eigen::MatrixXi blueDownsampled = downsampleandround(b, downsamplingFactor);
+
+	std::function<float(float)> is_not_nan_functor = [](float f) {return !std::isnan(f); };
+	std::function<float(float)> nan_to_zero_functor = nantozero;
+
+	Eigen::MatrixXf contrastNulled = (z.unaryExpr(is_not_nan_functor)).cwiseProduct(contrast.unaryExpr(nan_to_zero_functor));
+
+	auto contrastWeight = gridsum(contrastNulled, downsamplingFactor);
+
+	Eigen::MatrixXf xContrasted = x.cwiseProduct(contrastNulled);
+	Eigen::MatrixXf yContrasted = y.cwiseProduct(contrastNulled);
+	Eigen::MatrixXf zContrasted = z.cwiseProduct(contrastNulled);
+	Eigen::MatrixXf contrastContrasted = contrast.cwiseProduct(contrastNulled);
+
+	Eigen::MatrixXf xDownsampled = gridsum(xContrasted, downsamplingFactor).cwiseQuotient(contrastWeight);
+	Eigen::MatrixXf yDownsampled = gridsum(yContrasted, downsamplingFactor).cwiseQuotient(contrastWeight);
+	Eigen::MatrixXf zDownsampled = gridsum(zContrasted, downsamplingFactor).cwiseQuotient(contrastWeight);
+	Eigen::MatrixXf contrastDownsampled = gridsum(contrastContrasted, downsamplingFactor).cwiseQuotient(contrastWeight);
+
+	Zivid::PointCloud pointCloudDownsampled(redDownsampled.rows(), redDownsampled.cols());
+
+	for (int i = 0; i < redDownsampled.rows(); i++)
+	{
+		for (int j = 0; j < redDownsampled.cols(); j++)
+		{
+			pointCloudDownsampled(i, j).setRgb(redDownsampled(i, j), greenDownsampled(i, j), blueDownsampled(i, j));
+			pointCloudDownsampled(i, j).setContrast(contrastDownsampled(i, j));
+			pointCloudDownsampled(i, j).x = xDownsampled(i, j);
+			pointCloudDownsampled(i, j).y = yDownsampled(i, j);
+			pointCloudDownsampled(i, j).z = zDownsampled(i, j);
+		}
+	}
+
+	return pointCloudDownsampled;
+}
+
+Eigen::MatrixXi downsampleandround(const Eigen::MatrixXi& matrixi, int downsamplingFactor)
+{
+	Eigen::MatrixXf matrixf = matrixi.template cast<float>();
+
+	std::function<float(float)> round_wrap = [](float f) {return std::round(f); };
```

Comment:
Nitpick, call this `functor` instead of `wrap`, since we did that for the other ones.

Reply:
done.